tests/unit/test_upgrade.py

```diff
@@ -76,6 +76,31 @@ def test_build_upgrade_stack(harness):
     assert len(stack) == len(set(stack))
 
 
+@pytest.mark.parametrize("upgrade_state", ("idle", "ready"))
+@pytest.mark.parametrize("upgrade_stack", ([], [0]))
+def test_run_password_rotation_while_upgrading(harness, upgrade_state, upgrade_stack):
+    harness.charm.upgrade.peer_relation.data[harness.charm.unit].update({"state": upgrade_state})
+    harness.charm.upgrade.upgrade_stack = upgrade_stack
+    harness.set_leader(True)
+
+    mock_event = MagicMock()
+    mock_event.params = {"username": "admin"}
+
+    with (
+        patch("charm.KafkaCharm.healthy", new_callable=PropertyMock, return_value=True),
+        patch("auth.KafkaAuth.add_user"),
+    ):
+        harness.charm._set_password_action(mock_event)
+
+    if (not upgrade_stack) and (upgrade_state == "idle"):
```

Comment:
**todo:** I wouldn't do this check here, as it's too close to the actual implementation.

Perhaps add a `@pytest.mark.parametrize("passes", ("True", "False")` (or equivalent), and check for that.

Reply:
I don't think this is too close to the implementation. I mean: I'm not checking for what I log out, I'm checking for the actual UX. The use-case is that if you try to run password rotation while upgrading, you should fail and with a well-defined message, that we check for